rocksdb/src/moved.rs

```diff
@@ -0,0 +1,108 @@
+use {
+    crate::RocksEthTrieDb,
+    eth_trie::{EthTrie, Trie, DB},
+    move_binary_format::errors::PartialVMError,
+    move_core_types::{effects::ChangeSet, resolver::MoveResolver},
+    move_table_extension::{TableChangeSet, TableResolver},
+    moved::{
+        primitives::{KeyHashable, B256},
+        state_actor::HistoricResolver,
+        storage::{State, ToTreeValues},
+    },
+    std::sync::Arc,
+};
+
+/// A blockchain state implementation backed by [`rocksdb`] as its persistent storage engine.
+pub struct RocksDbState<'db> {
+    db: Arc<RocksEthTrieDb<'db>>,
+    resolver: HistoricResolver<RocksEthTrieDb<'db>>,
+    state_root: B256,
+}
+
+impl<'db> RocksDbState<'db> {
+    const STATE_ROOT_KEY: &'static str = "state_root";
+
+    pub fn new(db: Arc<RocksEthTrieDb<'db>>) -> Self {
+        let state_root = db
+            .db()
+            .get(Self::STATE_ROOT_KEY)
+            .unwrap()
+            .map(|v| B256::from_slice(&v))
+            .unwrap_or(B256::ZERO);
+
+        Self {
+            resolver: HistoricResolver::new(db.clone(), state_root),
+            state_root,
+            db,
+        }
+    }
+
+    fn persist_state_root(&self) -> Result<(), rocksdb::Error> {
+        self.db.db().put(Self::STATE_ROOT_KEY, self.state_root.0)
+    }
+
+    fn tree(&self) -> EthTrie<RocksEthTrieDb<'db>> {
+        let db = self.db.clone();
+
+        match self.state_root {
+            B256::ZERO => EthTrie::new(db),
+            root => EthTrie::from(db, root).unwrap(),
+        }
+    }
+
+    fn insert_change_set_into_merkle_trie(
```

Comment:
Maybe we should call this fn `persist_state_change` to keep it consistent with `persist_state_root`. Plus the insertion into Merkle trie is implied.

Reply:
This function looks nearly identical to the one for the in-memory implementation. I wonder if we can write one implementation shared by both rocksdb and in-mem by having it take the `EthTrie` as a parameter instead of `self`.